src/colmap/util/threading_ctrl.h

```diff
@@ -0,0 +1,259 @@
+// Copyright (c) 2023, ETH Zurich and UNC Chapel Hill.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+//     * Neither the name of ETH Zurich and UNC Chapel Hill nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include "colmap/util/logging.h"
+#include "colmap/util/timer.h"
+
+#include <atomic>
+#include <climits>
+#include <functional>
+#include <future>
+#include <list>
+#include <queue>
+#include <thread>
+#include <type_traits>
+#include <unordered_map>
+
+namespace colmap {
+
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wkeyword-macro"
+#endif
+
+#ifdef __clang__
+#pragma clang diagnostic pop  // -Wkeyword-macro
+#endif
+
+// Reimplementation of threading with thread-related functions outside
+// controller Following util/threading.h
+// BaseController that supports templating in ControllerThread
+class BaseController {
+ public:
+  BaseController();
+  virtual ~BaseController() = default;
+
+  enum {
+    STARTED_CALLBACK = INT_MIN,
+    FINISHED_CALLBACK,
+
+    // thread-related callbacks from here, only useful in the Thread class,
+    // empty as default
+    LOCK_MUTEX_CALLBACK,
+    BLOCK_IF_PAUSED_CALLBACK,
+    SIGNAL_SETUP_CALLBACK,
+    CHECK_VALID_SETUP_CALLBACK,
+  };
```

Comment:
Can't this go into the `ControllerThread` instead?

Reply:
Not needed after the new design